lib/media/preload_manager.js

```diff
@@ -724,37 +713,90 @@ shaka.media.PreloadManager = class extends shaka.util.FakeEventTarget {
+    const playableVariants = shaka.util.StreamUtils.getPlayableVariants(
+        this.manifest_.variants);
+    const adaptationSet = this.currentAdaptationSetCriteria_.create(
+        playableVariants);
+    // Guess what the first variant will be, based on a SimpleAbrManager.
+    this.abrManager_.configure(this.config_.abr);
+    this.abrManager_.setVariants(Array.from(adaptationSet.values()));
+
+    if (this.shouldCreateSegmentIndexBeforeDrmEngineInitialization_()) {
       const variant = this.abrManager_.chooseVariant();
       if (variant) {
-        const promises = [];
-        this.prefetchedVariant_ = variant;
-        if (variant.video) {
-          promises.push(this.prefetchStream_(variant.video, isLive));
-        }
-        if (variant.audio) {
-          promises.push(this.prefetchStream_(variant.audio, isLive));
+        const createSegmentIndexPromises = [];
+        for (const stream of [variant.video, variant.audio]) {
+          if (stream && !stream.segmentIndex) {
+            createSegmentIndexPromises.push(stream.createSegmentIndex());
+          }
         }
-        const textStream = this.chooseTextStream_();
-        if (textStream && shaka.util.StreamUtils.shouldInitiallyShowText(
-            variant.audio, textStream, this.config_)) {
-          promises.push(this.prefetchStream_(textStream, isLive));
-          this.prefetchedTextStream_ = textStream;
+        if (createSegmentIndexPromises.length > 0) {
+          await Promise.all(createSegmentIndexPromises);
         }
+      }
+    }
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  shouldCreateSegmentIndexBeforeDrmEngineInitialization_() {
+    goog.asserts.assert(
+        this.manifest_, 'The manifest should already be parsed.');
+
+    // If we only have one variant, it is useful to preload it, because it will
+    // be the only one we can use.
+    if (this.manifest_.variants.length == 1) {
+      return true;
+    }
+
+    // In HLS, DRM information is usually included in the media playlist, so we
+    // need to download the media playlist to get the real information.
+    if (this.manifest_.type == shaka.media.ManifestParser.HLS) {
+      return true;
+    }
 
-        await Promise.all(promises);
+    return false;
+  }
+
+  /**
+   * Prefetches segments.
+   *
+   * @return {!Promise}
+   * @private
+   */
+  async prefetchInner_() {
+    goog.asserts.assert(this.currentAdaptationSetCriteria_,
+        'Must have an AdaptationSetCriteria');
+
+    const playableVariants = shaka.util.StreamUtils.getPlayableVariants(
+        this.manifest_.variants);
+    const adaptationSet = this.currentAdaptationSetCriteria_.create(
+        playableVariants);
+    // Guess what the first variant will be, based on a SimpleAbrManager.
+    this.abrManager_.configure(this.config_.abr);
+    this.abrManager_.setVariants(Array.from(adaptationSet.values()));
+
+    const variant = this.abrManager_.chooseVariant();
```

Comment:
code duplication, pls move it to helper method

Reply:
Done!